xmipptomo/protocols/protocol_coords_roi.py

```diff
@@ -26,68 +26,112 @@
-        form.addParam('points', IntParam, label="Number of points", condition='selection == 0', allowsNull=True,
-                      help='see "Selection" help')
-        form.addParam('distance', IntParam, label='Distance', default=0,
-                      help='Maximum radial distance (in pixels) between mesh vertex and a coordinate to consider that '
-                           'it belongs to the ROI. Wizard returns three times the box size of the input coordinates.')
+                      help='Selection options:\n*Connected component*: It takes the whole connected component (cc) if '
+                           'a percentage of the points (introduced in the next field) in the cc belongs to the ROI. '
+                           '\n*Points*: It takes just the points of the cc which belongs to the roi')
+        form.addParam('points', IntParam, label="Percentage of coordinates in ROI",  default=80,
+                      condition='selection == 0',
+                      allowsNull=True, help='Percentage of coordinates from a connected component that should be inside'
+                                            ' the ROI to consider that connected component.')
+        form.addParam('distance', IntParam, label='Distance', default=50,
+                      help='Maximum euclidean distance (in pixels) between ROI vertex and a coordinate to consider that'
+                           ' it belongs to the ROI.')
 
     # --------------------------- INSERT steps functions --------------------------------------------
     def _insertAllSteps(self):
         self._insertFunctionStep('computeDistances')
 
     # --------------------------- STEPS functions -------------------------------
     def computeDistances(self):
-        inputCoor = self.inputCoordinates.get()
-        inputMesh = self.inputMesh.get()  # for REAL MESHES, inputMesh.getMesh() + .iterItems()???
-        distance = self.distance.get()
-        outputSet = self._createSetOfCoordinates3D(inputCoor.getPrecedents())
-        outputSet.copyInfo(inputCoor)
-        outputSet.setBoxSize(inputCoor.getBoxSize())
-        for coorm in inputMesh.iterItems():  # TODO: This works if mesh is a SetOfCoord
-            for coorc in inputCoor.iterItems():
-                if abs(coorm.getX() - coorc.getX()) <= distance and abs(coorm.getY() - coorc.getY()) <= distance \
-                        and abs(coorm.getZ() - coorc.getZ()) <= distance:
-                    outputSet.append(coorc)
-        # selection options??
-        self._defineOutputs(outputCoordinates=outputSet)
-        self._defineSourceRelation(inputCoor, outputSet)
+        sel = self.selection.get()
+        for ix, inputSetCoor in enumerate(self.inputCoordinates):
+            perc = self._percentage(inputSetCoor)
+            for mesh in self.inputMeshes.get().iterItems():
+                if inputSetCoor.get().getPrecedents().getFirstItem().getFileName() == mesh.getVolume().getFileName():
```

Comment:
This if look suspicious, why the getPrecedents().getFirstItem() here is so relevant?